python/gauss2d/fit/parameters.h

```diff
@@ -0,0 +1,211 @@
+    .def("__repr__", &Base::str);
+
+    py::class_<Class, std::shared_ptr<Class>, Name, Base>(m, pyclass_name.c_str())
+    .def(py::init<T, T, const std::string>(),
+         "min"_a=-std::numeric_limits<T>::infinity(),
+         "max"_a=std::numeric_limits<T>::infinity(),
+         "name"_a=""
+    )
+    .def("check", &Class::check)
+    .def("clip", &Class::clip)
+    .def_property("min", &Class::get_min, &Class::set_min)
+    .def_property("max", &Class::get_max, &Class::set_max)
+    .def("__repr__", &Class::str);
+}
+
+template<typename T, class C>
+auto declare_parameter_class(py::module &m, std::string name, std::string suffix=g2f::suffix_type_str<T>()) {
+    using Base = parameters::ParameterBase<T>;
+    std::string pyclass_name = name + "Parameter" + g2f::suffix_type_str<T>();
+    return py::class_<C, std::shared_ptr<C>, Base>(m, pyclass_name.c_str());
+}
+
+/*
+    Class is the CRTP class for CRTP parameters, whereas C is the "derived" class.
+    For classes derived from a CRTP parameter, the two are the same.
+    TODO: Come up with a better naming scheme for this.
+*/
+template<class Class, class C, typename... Args>
+auto declare_parameter_methods(py::class_<C, Args...> c) {
+    return c
+    .def_property_readonly("default",  &Class::get_default)
+    .def_property_readonly("desc", &Class::get_desc)
+    .def_property("fixed", &Class::get_fixed, &Class::set_fixed)
+    .def_property("free", &Class::get_free, &Class::set_free)
+    .def_property("inheritors", &Class::get_inheritors, &Class::set_inheritors)
+    .def_property("label", &Class::get_label, &Class::set_label)
+    .def_property("limits", &Class::get_limits, &Class::set_limits, py::keep_alive<1, 2>())
+    .def_property_readonly("linear", &Class::get_linear)
+    .def_property("modifiers", &Class::get_modifiers, &Class::set_modifiers)
+    .def_property_readonly("min", &Class::get_min)
+    .def_property_readonly("max", &Class::get_max)
+    .def_property_readonly("name", &Class::get_name)
+    .def_property_readonly("ptr", &Class::ptr)
+    // TODO: Figure out if it's possible to bind these
+    //.def_property_readonly_static("limits_maximal", &Class::limits_maximal)
+    .def_property("transform", &Class::get_transform, &Class::set_transform)
+    .def_property_readonly("transform_derivative", &Class::get_transform_derivative)
+    //.def_property_readonly_static("transform_none", &Class::transform_none)
+    .def_property("value", &Class::get_value, &Class::set_value)
+    .def_property("value_transformed", &Class::get_value_transformed, &Class::set_value_transformed)
+
+/*
+    Additional methods that might be worth wrapping.
+
+    inline std::shared_ptr<Limits<T>> & _get_limits() { return _limits_ptr; }
+    inline std::shared_ptr<Transform<T>> & _get_transform() { return _transform_ptr; }
+*/
+//    This doesn't work - should it?
+//    .def(py::self == py::self)
+//    .def(py::self != py::self)
+//    .def(hash(py::self))
```

Comment:
I'm not sure, what are you trying to do here?

Reply:
A lot of the C++ classes have comparison operators that are simply comparing addresses, so that they can be stored in STL containers like sets (as opposed to, say, having parameters compare values and units, which might be convenient for some contexts but would break set storage). Those lines are shorthand for overriding `__eq__`, `__ne__` and `__hash__`, but I couldn't get them to work (possibly due to [this bug](https://github.com/pybind/pybind11/issues/2191)) and anyway I don't think I need to override the Python default, which should effectively be the same (presumably `id(self) == id(other)`).